### PR TITLE
Allow arbitrary types of node ids for node2vec

### LINKED ISSUE #738: [BUG] Using layout_umap and getting key error
## Expected Behavior
a graph layout using UMAP for dim reduction to 2d

## Actual Behavior
coming from layouts\auto.py
 KeyError: '2'

## Example Code
Please see [How to create a Minimal, Reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) for some guidance on creating the best possible example of the problem
```python
import networkx as nx
from graspologic.layouts import layout_umap

g = nx.erdos_renyi_graph(10, 0.7, directed=False)
layout_umap(g)
```

## Full Traceback
```pytb
Paste the full traceback in case there is an exception

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-17-1c062a6ca19a> in <module>
      3 
      4 g = nx.erdos_renyi_graph(10, 0.7, directed=False)
----> 5 layout_umap(g)

~\miniconda3\lib\site-packages\graspologic\layouts\auto.py in layout_umap(graph, min_dist, n_neighbors, max_edges, random_seed)
    177         min_dist=min_dist, n_neighbors=n_neighbors, random_state=random_seed
    178     ).fit_transform(tensors)
--> 179     positions = _node_positions_from(lcc_graph, labels, points, random_seed=random_seed)
    180     return lcc_graph, positions
    181 

~\miniconda3\lib\site-packages\graspologic\layouts\auto.py in _node_positions_from(graph, labels, down_projection_2d, random_seed)
    241     scaled_points = _scale_points(down_projection_2d, covered_area)
    242     partitions = leiden(graph, random_seed=random_seed)
--> 243     positions = [
    244         NodePosition(
    245             node_id=key,

~\miniconda3\lib\site-packages\graspologic\layouts\auto.py in <listcomp>(.0)
    246             x=scaled_points[index][0],
    247             y=scaled_points[index][1],
--> 248             size=sizes[key],
    249             community=partitions[key],
    250         )

KeyError: '2'

```

## Your Environment
* Python version:  3.8
* graspologic version:

### PR DESCRIPTION
**Pull request by Nyecarr**
- [ ] Does this PR add any new dependencies?
- [x] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [x] Have you built the documentation (reference and/or tutorial) and verified the generated documentation is appropriate?
    - I have run the sphinx-build command specified in the makefile but I am not seeing my docstring changes reflected in the built docs -- send help.

#### Reference Issues/PRs
Fixes #738 

#### What does this implement/fix? Briefly explain your changes.
Original implementation of node2vec only truly worked if a graph's node ids were strings. Because gensim's word2vec expects strings to generate the random walks, we were getting incorrect results when passing in a graph with node ids of any type other than string. This PR will fix this issue by immediately creating a new graph that has the same exact edgelist except all node ids are remapped to a unique string. Then, before returning the embedding, we remap the results from word2vec back to the original node ids.

#### Any other comments?
N2v is used in when generating layouts so I added an additional test around layout_umap to verify both string and int node ids return as expected.

### CODE REVIEW

File: graspologic/utils/utils.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 from collections import Iterable
 from functools import reduce
 from pathlib import Path
-from typing import Any, List, Optional, Tuple, Union
+from typing import Any, Dict, List, Tuple, Union
```
Comment (Nyecarr) [inline graspologic/utils/utils.py:8]: Unused optional

File: graspologic/embed/n2v.py [hunk 2]
```diff
@@ -233,7 +238,13 @@ def __init__(
         inout_hyperparameter: float,
         random_state: Optional[np.random.RandomState] = None,
     ):
-        self.graph: nx.Graph = graph
+        self.original_graph: nx.Graph = graph
+
+        graph_with_new_ids, new_id_map = remap_node_ids(graph=graph)
```
Comment (Nyecarr) [inline graspologic/embed/n2v.py:243]: Create a representation of the graph where all original node ids are remapped to a new string node id, see utils.py

File: graspologic/embed/n2v.py [hunk 3]
```diff
@@ -473,7 +484,7 @@ def _alias_setup(probabilities: List[float]):
     """
     number_of_outcomes = len(probabilities)
     alias = np.zeros(number_of_outcomes)
-    sampled_probabilities = np.zeros(number_of_outcomes, dtype=np.int)
+    sampled_probabilities = np.zeros(number_of_outcomes, dtype=int)
```
Comment (Nyecarr) [inline graspologic/embed/n2v.py:487]: Getting a deprecated warning, fixed although irrelevant for this pr

File: graspologic/utils/__init__.py [hunk 4]
```diff
@@ -45,5 +46,6 @@
     "fit_plug_in_variance_estimator",
     "remove_vertices",
     "remap_labels",
+    "remap_node_ids",
     "to_weighted_edge_list",
```
Comment (Nyecarr) [inline graspologic/utils/__init__.py:50]: This is missing in the import list... didn't dig into it but I suspect there is an issue.

File: tests/embed/test_n2v.py [hunk 5]
```diff
@@ -1,25 +1,223 @@
 # Copyright (c) Microsoft Corporation and contributors.
```
Comment (Nyecarr) [inline tests/embed/test_n2v.py:1]: Combine 2 n2v test files. Correct testing file structure should follow module structure. Embedding tests belong in tests -> embed and not in the root.

File: tests/embed/test_n2v.py [hunk 6]
```diff
@@ -1,25 +1,223 @@
...
+import unittest
+
+from sklearn.metrics import adjusted_rand_score
+from sklearn.cluster import KMeans
+
+import graspologic as gc
+
+
+class Node2VecEmbedTest(unittest.TestCase):
+    def test_n2v_returns_same_labels_with_different_nodeid_types(self):
```
Comment (Nyecarr) [inline tests/embed/test_n2v.py:16]: This is the relevant test. Failing before code change, passing after.

File: tests/layouts/test_auto.py [hunk 7]
```diff
@@ -15,6 +17,34 @@ def test_get_bounds(self):
         self.assertEqual(10, maxx)
         self.assertEqual(5, maxy)
 
+    def test_layout_umap_string_node_ids(self):
```
Comment (Nyecarr) [inline tests/layouts/test_auto.py:20]: As this is a dependency of n2v, added explicit test to verify #738 is fixed.

File: graspologic/utils/__init__.py (hunk 4 above)
Comment (Nyecarr) [reply graspologic/utils/__init__.py:50]: `to_weighted_edge_list` to be clear

File: tests/embed/test_n2v.py [hunk 8]
```diff
@@ -1,25 +1,225 @@
...
+        w = g._get_walk_length_interpolated(
+            degree=5, percentiles=[2, 3, 4, 5, 6, 7, 8, 9], max_walk_length=10
+        )
+
+        self.assertEqual(w, expected_walk_length)
+
 
-if __name__ == "__main__":
-    unittest.main()
+_edge_list = """
```
Comment (dwaynepryce) [inline tests/embed/test_n2v.py:146]: we have a test utils and test_data package and you could stash this in there if you wanted.  I'd prefer that over adding it inline like this

Comment (Nyecarr) [reply tests/embed/test_n2v.py:146]: I can move it - this change is a result from the merging the other test file, not an added code change for this pr.

File: graspologic/utils/__init__.py (hunk 4 above)
Comment (dwaynepryce) [reply graspologic/utils/__init__.py:50]: to_weighted_edge_list should have been removed in a PR that merged to dev a while ago.  when did you cut this branch?

Comment (dwaynepryce) [reply graspologic/utils/__init__.py:50]: augh I missed the string representation in `__all__`: https://github.com/microsoft/graspologic/blob/dev/graspologic/utils/__init__.py#L48

Can you delete it in this PR?  it shouldn't be there.

Comment (Nyecarr) [reply graspologic/utils/__init__.py:50]: Sure

File: graspologic/embed/n2v.py [hunk 9]
```diff
@@ -128,7 +130,13 @@ def node2vec_embed(
         f"Completed. Ending time is {str(end)} Elapsed time is {str(start - end)}"
     )
 
-    return model.wv.vectors, model.wv.index2word
+    labels = node2vec_graph.original_graph.nodes()
+    remapped_labels = node2vec_graph.label_map_to_string
+
+    return (
+        np.array([model.wv.get_vector(remapped_labels[node]) for node in labels]),
+        np.array(labels),
```
Comment (dwaynepryce) [inline graspologic/embed/n2v.py:138]: should we np.array this and not just [] it? the reason I bring it up is technically networkx is going to be fine with you mixing the node types (ints and strings), whereas I'm not sure what np.array does if you send it a list of different types?

Then again .nodes() may be returning some sort of AtlasView object or something and numpy may be fine saying the type is of AltasView or something?

Comment (Nyecarr) [reply graspologic/embed/n2v.py]: Modified to return lists instead. No real reason to return np.arrays, was just trying to maintain the current api.

File: graspologic/embed/n2v.py [hunk 10]
```diff
@@ -22,7 +24,7 @@ def node2vec_embed(
     iterations: int = 1,
     interpolate_walk_lengths_by_node_degree: bool = True,
     random_seed: Optional[int] = None,
-) -> Tuple[np.ndarray, np.ndarray]:
+) -> Tuple[np.array, List]:
```
Comment (dwaynepryce) [inline graspologic/embed/n2v.py:27]: List[Any]

**Review by dwaynepryce — APPROVED** [review]

**Pull request merged**